Restructure conditional access handling in the control flow graph

333fred opened the pull request:
We had a bug in the CFG builder that caused regions for the conditional access temp to be unnecessarily extended past liveliness. To fix this, I've restructured the visit order to visit the instance only when actually accessed, which fixes this issue. I've also added some testing to demonstrate the handling around dropped NoneOperations in XML element accesses, as those need some special logic to ensure we aren't dropping nodes.

333fred commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -1110,6 +1110,13 @@ bool isLongLivedCaptureReferenceSyntax(SyntaxNode captureReferenceSyntax)
                                     break;
+
+                                case CSharp.SyntaxKind.ConditionalAccessExpression:
+                                    if (((CSharp.Syntax.ConditionalAccessExpressionSyntax)syntax.Parent).Expression == syntax)
```
>if (((CSharp.Syntax.ConditionalAccessExpressionSyntax)syntax.Parent).Expression == syntax) [](start = 36, length = 90)

AlekseyTs said:

> I think we should see if we want to fix the builder instead. Looking at the added `NestedConditionalAccess` unit-test, the life-time of capture 0 is extended unnecessarily.
> In any case I do not think we should be treating it as long lived (i.e. something that is known to span multiple statements). So the relaxation should probably be applied elsewhere

333fred commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IConditionalAccessExpression.cs:
```diff
@@ -1236,5 +1236,127 @@ void M1(C input1, C input2, C input3)
+    .locals {R2}
+    {
+        CaptureIds: [0] [2] [4]
+        Block[B1] - Block
```
>Block[B1] - Block [](start = 8, length = 17)

AlekseyTs said:

> I think we actually should have a nested region here. it should enclose B1, B2 and own capture [0].

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionPerfTests.cs:
```diff
@@ -350,7 +350,7 @@ static void F(IEnumerable<int[]> x)
         }
 
-        [Fact]
+        [ConditionalFact(typeof(NoIOperationValidation), Reason = "Timeouts")]
```
AlekseyTs said:

> Should we have an issue to track disable tests like that?

333fred replied:
I don't think that's particularly useful, as I'm not currently thinking that we'll be attempting to renable them in the future.

---
In reply to: [571246466](https://github.com/dotnet/roslyn/pull/51032#discussion_r571246466) [](ancestors = 571246466)

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for review. Like my previous porting prs, the original commit is prefixed with `Ported` and the new one is prefixed with `Fixup` for diffability.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for reviews.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.ConditionalAccessOperationTracker.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        private readonly struct ConditionalAccessOperationTracker
+        {
+            public readonly ArrayBuilder<IOperation>? Operations;
```
>public readonly ArrayBuilder<IOperation>? Operations; [](start = 12, length = 53)

Consider adding a comment what these operations are and how they are related to conditional access.

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3328,14 +3292,85 @@ private IOperation CallNullableMember(IOperation value, SpecialMember nullableMe
+            return VisitConditionalAccessTestExpression(testExpression);
+        }
+
+        static bool IsConditionalAccessInstancePresentInChildren(IOperation operation)
```
Note to self: Add modifier.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3192,79 +3191,52 @@ private IOperation CallNullableMember(IOperation value, SpecialMember nullableMe
+                    //                      See Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator.UnitTests.ExpressionCompilerTests.ConditionalAccessExpressionType
+                    //                      Because of this, the recursion to visit the child operations will never occur if we visit the WhenNull of the current
+                    //                      conditional access, so we need to manually visit the Operation of the conditional access now.
+                    testExpression = VisitConditionalAccessTestExpression(currentConditionalAccess.Operation);
```
>testExpression = VisitConditionalAccessTestExpression(currentConditionalAccess.Operation); [](start = 20, length = 90)

It look like an initialization of  ```testExpression``` with anything other than ```currentConditionalAccess.Operation``` is unexpected. For example, we do the following test below ```operation.Operation == testExpression```.

333fred replied:
The updates to `ConditionalAccess_11` and `ConditionalAccess_12` will reflect addressing this feedback.

---
In reply to: [573174786](https://github.com/dotnet/roslyn/pull/51032#discussion_r573174786) [](ancestors = 573174786)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3328,14 +3292,85 @@ private IOperation CallNullableMember(IOperation value, SpecialMember nullableMe
+
+        static bool IsConditionalAccessInstancePresentInChildren(IOperation operation)
+        {
+            if (operation is not NoneOperation)
```
>if (operation is not NoneOperation) [](start = 12, length = 35)

I think we should have a comment explaining rational behind this condition.

AlekseyTs replied:
If this is just an optimization, I am not sure if it is worth it.

---
In reply to: [573251761](https://github.com/dotnet/roslyn/pull/51032#discussion_r573251761) [](ancestors = 573251761)

333fred replied:
This was not an optimization. It was, however, a bug. So I've fixed that, and the updates to `ConditionalAccess_13` will reflect this.

---
In reply to: [573401448](https://github.com/dotnet/roslyn/pull/51032#discussion_r573401448) [](ancestors = 573401448,573251761)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3328,14 +3292,85 @@ private IOperation CallNullableMember(IOperation value, SpecialMember nullableMe
+
+            if (_currentConditionalAccessTracker.Operations.Count > 0 && !IsConditionalAccessInstancePresentInChildren(testExpression))
+            {
+                _ = VisitConditionalAccessTestExpression(_currentConditionalAccessTracker.Operations.Pop());
```
>_ = VisitConditionalAccessTestExpression(_currentConditionalAccessTracker.Operations.Pop()); [](start = 16, length = 92)

Is this code path reachable? It feels like similar check in ```VisitConditionalAccess``` should prevent us from getting here.

333fred replied:
I had thought so too, but I added it after `ConditionalAccessFlow_13` proved it was needed. Here's the original operation tree for the code (note the last conditional access with 2 none operation children):

```
IBlockOperation (3 statements) (OperationKind.Block, Type: null, IsInvalid) (Syntax: 'Sub M(c As  ... End Sub')
  IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsInvalid) (Syntax: 'result = c?.X?.@a1?.<e1>')
    Expression: 
      ISimpleAssignmentOperation (OperationKind.SimpleAssignment, Type: System.Object, IsInvalid, IsImplicit) (Syntax: 'result = c?.X?.@a1?.<e1>')
        Left: 
          IParameterReferenceOperation: result (OperationKind.ParameterReference, Type: System.Object) (Syntax: 'result')
        Right: 
          IConversionOperation (TryCast: False, Unchecked) (OperationKind.Conversion, Type: System.Object, IsInvalid, IsImplicit) (Syntax: 'c?.X?.@a1?.<e1>')
            Conversion: CommonConversion (Exists: False, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
            Operand: 
              IConditionalAccessOperation (OperationKind.ConditionalAccess, Type: ?, IsInvalid) (Syntax: 'c?.X?.@a1?.<e1>')
                Operation: 
                  IParameterReferenceOperation: c (OperationKind.ParameterReference, Type: C) (Syntax: 'c')
                WhenNotNull: 
                  IConditionalAccessOperation (OperationKind.ConditionalAccess, Type: ?, IsInvalid) (Syntax: '.X?.@a1?.<e1>')
                    Operation: 
                      IFieldReferenceOperation: C.X As System.Xml.Linq.XElement (OperationKind.FieldReference, Type: System.Xml.Linq.XElement) (Syntax: '.X')
                        Instance Receiver: 
                          IConditionalAccessInstanceOperation (OperationKind.ConditionalAccessInstance, Type: C, IsImplicit) (Syntax: 'c')
                    WhenNotNull: 
                      IConditionalAccessOperation (OperationKind.ConditionalAccess, Type: ?, IsInvalid) (Syntax: '.@a1?.<e1>')
                        Operation: 
                          IOperation:  (OperationKind.None, Type: null) (Syntax: '.@a1')
                        WhenNotNull: 
                          IOperation:  (OperationKind.None, Type: null, IsInvalid) (Syntax: '.<e1>')
  ILabeledOperation (Label: exit) (OperationKind.Labeled, Type: null, IsImplicit) (Syntax: 'End Sub')
    Statement: 
      null
  IReturnOperation (OperationKind.Return, Type: null, IsImplicit) (Syntax: 'End Sub')
    ReturnedValue: 
      null
```

---
In reply to: [573257576](https://github.com/dotnet/roslyn/pull/51032#discussion_r573257576) [](ancestors = 573257576)

AlekseyTs replied:
It looks like there was a change in behavior reflected in ```ConditionalAccessFlow_13```, which was likely caused by the fix in ```IsConditionalAccessInstancePresentInChildren```. Do we still need this check here after the fix? Could you please elaborate why?

---
In reply to: [573954488](https://github.com/dotnet/roslyn/pull/51032#discussion_r573954488) [](ancestors = 573954488,573257576)

AlekseyTs commented:
Done with review pass (commit 3)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [776337916](https://github.com/dotnet/roslyn/pull/51032#issuecomment-776337916) [](ancestors = 776337916)

AlekseyTs commented:
Done with review pass (commit 4)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [777158691](https://github.com/dotnet/roslyn/pull/51032#issuecomment-777158691) [](ancestors = 777158691)

AlekseyTs reviewed (approved):
LGTM (commit 5), assuming CI is passing

333fred commented:
@dotnet/roslyn-compiler for a second review please.

333fred commented:
@dotnet/roslyn-compiler for a second review please.

chsienki commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3192,79 +3191,52 @@ private IOperation CallNullableMember(IOperation value, SpecialMember nullableMe
-
-            var whenNull = new BasicBlockBuilder(BasicBlockKind.Block);
-
             currentConditionalAccess = operation;
```
This re-assignment is no longer needed, right?

chsienki commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3192,79 +3191,52 @@ private IOperation CallNullableMember(IOperation value, SpecialMember nullableMe
             IOperation testExpression;
+            var whenNull = new BasicBlockBuilder(BasicBlockKind.Block);
+            var previousTracker = _currentConditionalAccessTracker;
+            _currentConditionalAccessTracker = new ConditionalAccessOperationTracker(operations, whenNull);
```
NIT: consider moving this below the while loop, to make it clearer how the operations are being mutated. It took me a bit to figure out how the stack was actually getting built.

333fred replied:
I can't. The loop needs this to be set up correctly in case it encounters a NoneOperation.

chsienki reviewed (approved):
:shipit:
